Decorrelate subqueries with Limit or TopN

kasiafi opened the pull request:
This change allows to decorrelate subqueries containing ORDER BY + LIMIT or LIMIT clause.
Refers to: https://github.com/prestosql/presto/issues/1358

kokosing commented on presto-main/src/test/java/io/prestosql/sql/query/TestSubqueries.java:
```diff
@@ -96,10 +96,9 @@ public void testCorrelatedSubqueriesWithLimit()
                 "VALUES 1");
-        // cannot enforce LIMIT on correlated subquery
-        assertions.assertFails(
+        assertions.assertQuery(
```
Can you also please test:

```                
SELECT (SELECT t.a FROM (VALUES 1, 1, 2, 3) t(a) WHERE t.a=t2.b LIMIT 2) from (VALUES 1) t2(b)
```

That should fail.

kokosing commented on presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PlanNodeDecorrelator.java:
```diff
@@ -155,22 +165,37 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
             }
 
-            Set<Symbol> constantSymbols = childDecorrelationResult.getConstantSymbols();
+            if (node.getCount() > 1) {
```
please invert the condition so methods declared below will in the same order as usage

kokosing commented on presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PlanNodeDecorrelator.java:
```diff
@@ -145,7 +155,7 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
         public Optional<DecorrelationResult> visitLimit(LimitNode node, Void context)
         {
-            if (node.getCount() == 0) {
+            if (node.getCount() == 0 || node.isWithTies()) {
```
can you please add a test for such queries?

kasiafi replied:
I needed to add the check to prevent wrong usage of this decorrelator. However, I can't create a test that would physically hit this path. This is because `LimitNode` `withTies` is rewritten before the decorrelation phase.

kasiafi replied:
I will assert that a correlated subquery containing `FETCH FIRST ... WITH TIES` fails (but this is due to WindowNode not decorrelating).

kokosing commented on presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PlanNodeDecorrelator.java:
```diff
@@ -155,22 +165,37 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
+            if (node.getCount() > 1) {
+                return rewriteLimitWithRowCountGreaterThanOne(childDecorrelationResult, node.getCount());
+            }
+            else {
```
extra `else` block

kokosing commented on presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PlanNodeDecorrelator.java:
```diff
@@ -155,22 +165,37 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
             }
 
-            if (node.getCount() != 1) {
+            if (childDecorrelationResult.getConstantSymbols().isEmpty()) {
```
how is this possible? Did we try to decorelate  uncorrelated subquery?

sopel39 replied:
Constant symbols might be empty when correlated predicates are of form: `corr_x > uncorr_x`

sopel39 replied:
Let's keep this check specific to `rewriteLimitWithRowCountGreaterThanOne` or `rewriteLimitWithRowCountOne` as this code is already tricky

kokosing commented on presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PlanNodeDecorrelator.java:
```diff
@@ -189,6 +214,80 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
+                    decorrelatedChildNode,
+                    ImmutableList.copyOf(childDecorrelationResult.getConstantSymbols()),
+                    symbolAllocator.newSymbol("row_number", BigintType.BIGINT),
+                    Optional.of(toIntExact(limit)),
```
maybe we should change `maxRowCountPerPartition` be an `Optional<Long>`?

kasiafi replied:
I think that the choice of Integer was intentional. It reflects the fact that:
`// Operators can handle MAX_VALUE rows per page ...` - as a comment in the class `WindowFilterPushDown` says.
It is also consistent with `TopNRowNumberNode` having `int maxRowCountPerPartition`.

kokosing commented on presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PlanNodeDecorrelator.java:
```diff
@@ -189,6 +214,80 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
+                    false,
+                    Optional.empty());
+
+            // remove row_number symbol
```
is this needed? Isn't this be removed anyway further by optimizer?

kokosing commented on presto-tests/src/main/java/io/prestosql/tests/AbstractTestQueries.java:
```diff
@@ -5173,6 +5173,61 @@ public void testLateralJoin()
                 "VALUES (1, 3)");
+
+        // Limit in correlated subquery
+        assertQuery(
```
please add this test to `TestSubqueries` (and it would be nice to move also above non-connector related subquery tests as well in separate commit)

kasiafi replied:
Out of the three tests I added here, the one using tables cannot be moved to `TestSubqueries`. Shall I leave that one here and move the other two, or is there a way to test queries using tables in `TestSubqueries`?

kokosing replied:
Yes, please.

kokosing commented on presto-tests/src/main/java/io/prestosql/tests/AbstractTestQueries.java:
```diff
@@ -5173,6 +5173,61 @@ public void testLateralJoin()
+                        "(1, 'a'), " +
+                        "(1, 'a'), " +
+                        "(1, 'a'), " +
+                        "(2, 'b')) inner_relation(f0, f1) " +
```
`fo, f1` -> `id, value`?

kokosing commented on presto-tests/src/main/java/io/prestosql/tests/AbstractTestQueries.java:
```diff
@@ -5173,6 +5173,61 @@ public void testLateralJoin()
+        // Limit in correlated subquery
+        assertQuery(
+                "SELECT * " +
+                        "FROM (VALUES 1, 2, 3) outer_relation(id) " +
```
can you add `null` to `id` column? same for inner relation.

kokosing commented on presto-tests/src/main/java/io/prestosql/tests/AbstractTestQueries.java:
```diff
@@ -5173,6 +5173,61 @@ public void testLateralJoin()
+                        "ORDER BY inner_relation.f1 LIMIT 2) " +
+                        "ON TRUE",
+                "VALUES " +
+                        "(1, 'b'), " +
```
Order is not respected after a join and so `b` is before `a`, so this test could be flaky as it possible to get different results. Please add `ORDER BY 1,2` at the end of your query.

kasiafi replied:
No, `assertQuery()` compares the results ignoring order, which is actually a must for testing any query without `ORDER BY` at top level.

kokosing commented on presto-tests/src/main/java/io/prestosql/tests/AbstractTestQueries.java:
```diff
@@ -5173,6 +5173,61 @@ public void testLateralJoin()
+                        "(4, 'EGYPT')");
+        assertQuery(
+                "SELECT * " +
+                        "FROM (VALUES 1, 2, 3) outer_relation(id) " +
```
add nulls as id column

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PlanNodeDecorrelator.java:
```diff
@@ -155,22 +165,37 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
+        }
+
+        // TODO Limit (1) could be decorrelated by the method rewriteLimitWithRowCountGreaterThanOne() as well.
+        // This would enable decorrelating more subqueries that it is possible now
```
Mention what is not supported ATM, e.g: limits 1 with ties

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PlanNodeDecorrelator.java:
```diff
@@ -155,22 +165,37 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
+        {
             PlanNode decorrelatedChildNode = childDecorrelationResult.node;
 
-            if (constantSymbols.isEmpty() || !constantSymbols.containsAll(decorrelatedChildNode.getOutputSymbols())) {
```
undo

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PlanNodeDecorrelator.java:
```diff
@@ -189,6 +214,80 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
+            RowNumberNode rowNumberNode = new RowNumberNode(
+                    idAllocator.getNextId(),
+                    decorrelatedChildNode,
+                    ImmutableList.copyOf(childDecorrelationResult.getConstantSymbols()),
```
This is wrong. Think about:
```
corr_x = a
corr_y < b
```
and `LIMIT n`
For each pair of `(corr_x, corr_y)` you want to have `n` rows, but if you just partition by `a` you will flatten such pairs.

I think you have to have condition that all `symbolsToPropagate` are also constants. Please also add test.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PlanNodeDecorrelator.java:
```diff
@@ -189,6 +214,80 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
+                    node.getId(),
+                    decorrelatedChildNode,
+                    new Specification(
+                            ImmutableList.copyOf(childDecorrelationResult.getConstantSymbols()),
```
similar case here

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PlanNodeDecorrelator.java:
```diff
@@ -189,6 +214,80 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
+
+            // rewrite Limit to RowNumberNode partitioned by constant symbols
+            RowNumberNode rowNumberNode = new RowNumberNode(
+                    idAllocator.getNextId(),
```
use id from the original node

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PlanNodeDecorrelator.java:
```diff
@@ -189,6 +214,80 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
+            OrderingScheme mappedOrderingScheme = childDecorrelationResult.getCorrelatedSymbolMapper().map(node.getOrderingScheme());
+
+            TopNRowNumberNode topNRowNumberNode = new TopNRowNumberNode(
+                    node.getId(),
```
use is from the original node.

kasiafi commented:
Applied comments.

kasiafi commented on presto-main/src/test/java/io/prestosql/sql/query/TestSubqueries.java:
```diff
@@ -118,6 +118,9 @@ public void testCorrelatedSubqueriesWithLimit()
                 "VALUES true",
                 false);
+        assertions.assertFails(
+                "SELECT (SELECT t.a FROM (VALUES 1, 2, 3) t(a) WHERE t.a=t2.b ORDER BY a FETCH FIRST ROW WITH TIES) FROM (VALUES 1) t2(b)",
```
I'll add spaces in the test cases I added. However, the whole file is formatted likewise (no spaces around `=`, `<` etc.).
I'd rather not fix it all in this PR.

kasiafi commented:
Applied.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PlanNodeDecorrelator.java:
```diff
@@ -189,6 +218,78 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
+            Set<Symbol> constantSymbols = childDecorrelationResult.getConstantSymbols();
+
+            if (constantSymbols.isEmpty() || !constantSymbols.containsAll(childDecorrelationResult.symbolsToPropagate)) {
+                return Optional.empty();
```
nit: in such cases we could rewrite to `RowNumberNode` above decorrelated join.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PlanNodeDecorrelator.java:
```diff
@@ -189,6 +218,78 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
+            Set<Symbol> constantSymbols = childDecorrelationResult.getConstantSymbols();
+
+            if (constantSymbols.isEmpty() || !constantSymbols.containsAll(childDecorrelationResult.symbolsToPropagate)) {
+                return Optional.empty();
```
nit: in such cases we could rewrite to `TopNRowNumberNode` above decorrelated join

kasiafi replied:
Yes, but it would also require assigning unique ids to input rows and adding them to `TopNRowNumberNode`'s partitioning symbols to avoid equal input rows sharing subquery results.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PlanNodeDecorrelator.java:
```diff
@@ -189,6 +218,78 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
+                    new Specification(
+                            ImmutableList.copyOf(childDecorrelationResult.symbolsToPropagate),
+                            Optional.of(mappedOrderingScheme)),
+                    symbolAllocator.newSymbol("row_number", BigintType.BIGINT),
```
static import

sopel39 commented on presto-main/src/test/java/io/prestosql/sql/query/TestSubqueries.java:
```diff
@@ -119,8 +137,51 @@ public void testCorrelatedSubqueriesWithLimit()
+                        "(2, 'w'), " +
+                        "(null, 'x')) inner_relation(id, value) " +
+                        "WHERE outer_relation.id = inner_relation.id " +
+                        "ORDER BY inner_relation.value LIMIT 2) " +
```
add test where ordering scheme is also correlated (e.g `ORDER BY outer_relation.id`)

sopel39 replied:
ping, you are still missing test for the case above.

sopel39 replied:
It's different case than "outer_relation.id = inner_relation.id"

It should be possible to write such test that passes.

kasiafi replied:
Ah, right. You meant  order by outer symbol that is in equality predicate? That won't decorrelate either, and for the same reason as the previous test. I could add a test but it won't bring anything new.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PlanNodeDecorrelator.java:
```diff
@@ -189,6 +218,78 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
+            PlanNode decorrelatedChildNode = childDecorrelationResult.node;
+            Set<Symbol> constantSymbols = childDecorrelationResult.getConstantSymbols();
+
+            if (constantSymbols.isEmpty() || !constantSymbols.containsAll(childDecorrelationResult.symbolsToPropagate)) {
```
AFAIR this condition was required in `limit -> aggregation` rewrite to avoid creating default aggregations. Maybe that conditions should be refactored?

kasiafi commented on presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PlanNodeDecorrelator.java:
```diff
@@ -189,6 +218,78 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
+        @Override
+        public Optional<DecorrelationResult> visitTopN(TopNNode node, Void context)
+        {
+            if (node.getCount() == 0) {
```
`LIMIT 0` is legal. Anyway it should be trimmed before decorrelation.

kokosing replied:
maybe you could use `checkState(node.getCount() !=  0, "LIMIT 0 should be trimmed before decorrelation")`?

martint replied:
> maybe you could use checkState(node.getCount() != 0, "LIMIT 0 should be trimmed before decorrelation")?

Since there's no guarantee rules are applied in any particular order, that would force the caller to do this check and skip calling this method, so not sure we gain anything. Since we already use `Optional.empty()` to signal "couldn't do anything", I'd keep it that way.

kasiafi commented:
Added support for the case when no local symbols are related to correlation symbols.
Added removing constant symbols from TopN's ordering scheme.

Planning to add more tests.

kasiafi commented:
Refactored, added more tests.
@sopel39 @kokosing please review.

sopel39 commented on presto-main/src/test/java/io/prestosql/sql/query/TestSubqueries.java:
```diff
@@ -119,8 +143,113 @@ public void testCorrelatedSubqueriesWithLimit()
+                        "FROM (VALUES 1, 2, 3, null) outer_relation(id) " +
+                        "LEFT JOIN LATERAL " +
+                        "(SELECT value FROM (VALUES 'c', 'a', 'b') inner_relation(value) " +
+                        "   WHERE outer_relation.id = 3 ORDER BY outer_relation.id LIMIT 2) " +
```
Shouldn't `decorrelateOrderingScheme` decorrelate `outer_relation.id`?

kasiafi replied:
The correlation symbol in `ORDER BY` is resolved; the query fails due to projection `outer_relation.id -> outer_relation.id` not decorrelating.

sopel39 replied:
I wonder why this projection appears on the first place. it's not relevant though.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PlanNodeDecorrelator.java:
```diff
@@ -189,6 +219,143 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
+            ImmutableList.Builder<Symbol> nonConstantOrderBy = ImmutableList.builder();
+            ImmutableMap.Builder<Symbol, SortOrder> nonConstantOrderings = ImmutableMap.builder();
+            for (Symbol symbol : orderingScheme.getOrderBy()) {
+                if (!constantSymbols.contains(symbol) && !correlation.contains(symbol)) {
```
Could you point me to actual query that test `!correlation.contains(symbol)`?

kasiafi replied:
That would be any query with `ORDER BY` a local symbol - the symbol passes the condition.
An example with a symbol that doesn't pass:
```
assertions.assertFails(
                "SELECT * " +
                        "FROM (VALUES 1, 2, 3, null) outer_relation(id) " +
                        "LEFT JOIN LATERAL " +
                        "(SELECT value FROM (VALUES 'c', 'a', 'b') inner_relation(value) " +
                        "   WHERE outer_relation.id = 3 ORDER BY outer_relation.id LIMIT 2) " +
                        "ON TRUE",
                UNSUPPORTED_CORRELATED_SUBQUERY_ERROR_MSG);
```
(I need to fix the comment for this one).

sopel39 replied:
I mean if we could have a test that shows improvement vs using mapper. I though the one with "corr_sym = 3 order by corr_sym" would be such test

kasiafi replied:
Yes, this one matches too.
The problem is that for any query that has ordering by a correlation symbol, there 'appears' an identity projection on this symbol below `TopNNode`.
And so far, `ProjectNode`'s assignments aren't decorrelated. So currently this path cannot be tested, because the query will fail with `Given correlated subquery not supported`, although `TopNNode` itself decorrelates.
That's why I intended to decorrelate Projections, as they are the real blocker.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PlanNodeDecorrelator.java:
```diff
@@ -189,6 +219,143 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
+                                node.getCount() == 1));
+                    })
+                    .orElseGet(() -> {
+                        // no ordering symbols are left - rewrite TopN to RowNumberNode partitioned by constant symbols
```
ditto: `TopNRowNumberNode` to `RowNumberNode` feels like it should be separate rule when ordering scheme is empty.

kasiafi replied:
`OrderingScheme` cannot be empty.

kasiafi commented:
@kokosing @sopel39 does it need changes?

kasiafi commented:
> I wonder why this projection appears on the first place. it's not relevant though.

the correlated symbol is introduced to the subquery plan by Planner in the form of projection `projected -> corr`. And `OrderingScheme` originally uses the `projected` symbol.
Afterwards, `UnaliasSymbolReferences` assimilates this projection and distributes the `corr` symbol through the plan (including the `OrderingScheme`). I think the identity projection `corr -> corr` is a leftover of that.

sopel39 commented:
> I think the identity projection corr -> corr is a leftover of that.

Strange that it's not removed by `RemoveRedundantIdentityProjections`, but maybe it's not identity

kasiafi commented on presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PlanNodeDecorrelator.java:
```diff
@@ -189,6 +219,143 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
+                                new Specification(
+                                        ImmutableList.copyOf(childDecorrelationResult.symbolsToPropagate),
+                                        Optional.of(orderingScheme)),
+                                symbolAllocator.newSymbol("row_number", BIGINT),
```
It is pruned right above `RowNumberNode` or `TopNRowNumberNode` by `PruneUnreferencedOutputs`.

bitsondatadev commented on presto-main/src/test/java/io/prestosql/sql/query/TestSubqueries.java:
```diff
@@ -96,10 +96,34 @@ public void testCorrelatedSubqueriesWithLimit()
+                "VALUES 1");
+        assertions.assertFails(
+                "SELECT (SELECT t.a FROM (VALUES 1, 1, 2, 3) t(a) WHERE t.a = t2.b LIMIT 2) FROM (VALUES 1) t2(b)",
+                "Scalar sub-query has returned multiple rows");
```
This exception comes from the execution. So this is planned and optimized, then executed and afterwards we find that more than one row returns so this fails.
